src/vs/editor/contrib/codeAction/browser/codeActionMenu.ts

```diff
@@ -305,28 +366,107 @@ export class CodeActionMenu extends Disposable implements IEditorContribution {
 		renderDisposables.add(this.codeActionList.value.onDidChangeSelection(e => this._onListSelection(e)));
 		renderDisposables.add(this._editor.onDidLayoutChange(e => this.hideCodeActionWidget()));
 
-		// Populating the list widget and tracking enabled options.
+		enum CodeActionArrayGroup {
+			quickfix = 0,
+			extract = 1,
+			convert = 2,
+			normalRefactor = 3,
+			separator = 4,
+			other = 5,
+		}
+
+		// Filters and groups code actions by their group
+		const menuEntries: IAction[][] = [[], [], [], [], [], []];
+		let groupID = 0;
 		inputArray.forEach((item, index) => {
+			if (item instanceof CodeActionAction) {
+				// create codea action kind from string
+				// not array, but class -> add into class
+				// TODO:
+				const optionKind = item.action.kind;
+
+				if (CodeActionKind.QuickFix.contains(new CodeActionKind(String(optionKind)))) {
```

Comment:
I'm able to use `CodeActionKind`  here because all quickfixes will be under the same `CodeActionKind`. In the case of refactorings, the only type available in `type.js` is `CodeActionKind.Refactor`, so Rewrites, Converts, and Extracts all fall under this.